lib/mongo/collection/view/change_stream.rb

```diff
@@ -0,0 +1,138 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+require 'mongo/collection/view/change_stream/retryable'
+
+module Mongo
+  class Collection
+    class View
+
+      # Provides behaviour around a `$changeStream` pipeline stage in the
+      # aggregation framework. Specifying this stage allows users to request that
+      # notifications are sent for all changes to a particular collection or database.
+      #
+      # @note Only available in server versions 3.6 and higher.
+      #
+      # @since 2.5.0
+      class ChangeStream < Aggregation
+        include Retryable
+
+        # @return [ String ] The fullDocument option default value.
+        #
+        # @since 2.5.0
+        FULL_DOCUMENT_DEFAULT = 'default'.freeze
+
+        # @return [ BSON::Document ] The change stream options.
+        #
+        # @since 2.5.0
+        attr_reader :options
+
+        # Initialize the change stream for the provided collection view, pipeline
+        # and options.
+        #
+        # @example Create the new change stream view.
+        #   ChangeStream.new(view, pipeline, options)
+        #
+        # @param [ Collection::View ] view The collection view.
+        # @param [ Array<Hash> ] pipeline The pipeline of operators to filter the change notifications.
+        # @param [ Hash ] opts The change stream options.
+        #
+        # @option options [ String ] :full_document Allowed values: ‘default’, ‘updateLookup’. Defaults to ‘default’.
+        #   When set to ‘updateLookup’, the change notification for partial updates will include both a delta
+        #   describing the changes to the document, as well as a copy of the entire document that was changed
+        #   from some time after the change occurred.
+        # @option options [ BSON::Document, Hash ] :resume_after Specifies the logical starting point for the
+        #   new change stream.
+        # @option options [ Integer ] :max_await_time_ms The maximum amount of time for the server to wait
+        #   on new documents to satisfy a change stream query.
+        # @option options [ Integer ] :batch_size The number of documents to return per batch.
+        # @option options [ BSON::Document, Hash ] :collation The collation to use.
+        #
+        # @since 2.5.0
+        def initialize(view, pipeline, options = {})
+          @view = view
+          @change_stream_filters = pipeline
+          @options = options.freeze
+          @resume_token = @options[:resume_after]
+          create_cursor!
```

Comment:
I thought I remember reading that the initial aggregation command should not be retryed but I don't see it in the spec. I'll open a spec question.

Reply:
Waiting for an answer: https://jira.mongodb.org/browse/SPEC-947

Reply:
Thanks for pointing this out. I chose to retry it because we historically retry read operations in the driver. 

([here](https://github.com/mongodb/mongo-ruby-driver/blob/master/lib/mongo/collection/view/iterable.rb#L39) is where we retry finds, aggregations, and [here](https://github.com/mongodb/mongo-ruby-driver/blob/master/lib/mongo/collection/view/readable.rb#L138) is where we retry a count, for example)

and I thought it would be inconsistent to not retry the initial aggregation for a change stream. I'll see how the discussion goes in the spec ticket, thanks for opening it.

Reply:
Thanks, this makes a lot more sense now that I know Ruby retries read operations already.